core/src/main/scala/cats/derived/traverse.scala

```diff
@@ -0,0 +1,121 @@
+package cats.derived
+
+import cats.derived.MkTraverse.SafeTraverse
+import cats.syntax.all._
+import cats.{Applicative, Eval, Monoid, Traverse}
+import shapeless._
+
+import scala.annotation.implicitNotFound
+
+/**
+  * This trait extends `Traverse` and implements it's fold methods.
+  * The overriden implementation can be removed as soon as issue
+  * <a href="https://github.com/typelevel/cats/issues/107#issuecomment-393797529">cats-107</a>
+  * gets resolved.
+  * The current implementation is based on a proposal posted by @Baccata in a comment to the issue.
+  *
+  */
+@implicitNotFound("Could not derive an instance of Traverse[${F}]")
+trait MkTraverse[F[_]] extends Traverse[F] {
+
+  override def traverse[G[_], A, B](fa: F[A])(f: A => G[B])(implicit evidence$1: Applicative[G]): G[F[B]] =
+    safeTraverse(fa)(a => Eval.now(f(a))).value
+
+  def safeTraverse[G[_] : Applicative, A, B](fa: F[A])(f: A => Eval[G[B]]): Eval[G[F[B]]]
+
+  override def foldMap[A, B: Monoid](fa: F[A])(f: A => B): B = {
+    traverse[cats.data.Const[B, ?], A, B](fa) { a => cats.data.Const(f(a)) } getConst
+  }
+
+  override def foldRight[A, B](fa: F[A], lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B] = {
+    foldMap(fa)(f.curried andThen defer)(fMonoid(_ compose _)).apply(lb)
+  }
+
+  override def foldLeft[A, B](fa: F[A], b: B)(f: (B, A) => B): B =
+    foldMap[A, B => B](fa) { a => b => f(b, a) }(fMonoid(_ andThen _)).apply(b)
+
+  private def fMonoid[A](c: (A => A, A => A) => (A => A)): Monoid[A => A] = new Monoid[A => A] {
+    def combine(f: A => A, g: A => A): A => A = c(f, g)
+
+    def empty: A => A = identity
+  }
+
+  private def defer[B](f: Eval[B] => Eval[B]): Eval[B] => Eval[B] =
+    evalB => Eval.defer(f(evalB))
+
+}
+
+object MkTraverse extends MkTraverseDerivation {
+  def apply[F[_]](implicit mff: MkTraverse[F]): MkTraverse[F] = mff
+
+  private[derived] implicit class SafeTraverse[F[_]](val F: Traverse[F]) extends AnyVal {
+    def safeTraverse[G[_] : Applicative, A, B](fa: F[A])(f: A => Eval[G[B]]): Eval[G[F[B]]] = F match {
+      case mk: MkTraverse[F] => mk.safeTraverse(fa)(f)
+      case _ => F.traverse[λ[t => Eval[G[t]]], A, B](fa)(f)(aEval[G])
+    }
+  }
+}
+
+trait MkTraverseDerivation extends MkTraverse0 {
+  implicit val mkTraverseId: MkTraverse[shapeless.Id] = new MkTraverse[shapeless.Id] {
+    override def safeTraverse[G[_] : Applicative, A, B](fa: Id[A])(f: A => Eval[G[B]]): Eval[G[Id[B]]] = f(fa)
+  }
+}
+
+trait MkTraverse0 extends MkTraverse1 {
+  // Induction step for products
+  implicit def mkTraverseHcons[F[_]](implicit ihc: IsHCons1[F, TraverseOrMk, MkTraverse]): MkTraverse[F] =
+    new MkTraverse[F] {
+      override def safeTraverse[G[_] : Applicative, A, B](fa: F[A])(f: A => Eval[G[B]]): Eval[G[F[B]]] = {
+        for {
+          ht <- Eval.now(ihc.unpack(fa))
+          th <- ihc.fh.unify.safeTraverse(ht._1)(f)
```

Comment:
Curiously, when implemented with `Applicative[Eval].compose[G]`, it was not stacksafe... The `Eval.flatMap` was necesseary. I need to dig deeper to know why.